[flutter_tools] Add support for custom devices

ardera opened the pull request:
- added a feature flag that allows enabling / disabling the custom devices feature
- add a config file in the default flutter config location (for windows, `%APPDATA%/.flutter_custom_devices.json`).
- the config file is created with the `$schema` tag and some (disabled) example device configured each time any flutter command is run when the custom devices feature is enabled.
- the `schema` tag points to `https://raw.githubusercontent.com/flutter/flutter/stable/packages/flutter_tools/static/custom-devices.schema.json` so we get autocompletion. (EDIT: Changed it to stable. Probably better that way, it'll probably confuse people if they get autocorrection for features that aren't yet implemented in stable. If people do want to get autocompletion for master channel features they can just change the URL)
- added a `CustomDevicesConfig` class for querying the currently configured custom devices (with `CustomDeviceConfig` representing a specific device config)
- added string interpolation to `utils.dart` (used for replacing things like `${hostPort}` in the custom device commands)
- added some testing
- did some manual testing and it seems at least the VS Code plugin needs the `linux-desktop` flag (plus the custom devices flag) enabled to even list any custom devices. not sure whats up with that
- also made protocol discovery not assert a loopback address since some people may not use port forwarding
- by default the target platform is `linux_arm64` (doesn't matter anyway since we're only building platform-agnostic stuff rn), didn't want to add a new target platform since that's probably a bigger breaking change
- there's a TODO item in the code to make the target platform configurable in the future

- resolves #65067

## Pre-launch Checklist
- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

~~It's probably better to add some more testing, for example for `CustomDevicePortForwarder`. Is there anything else I didn't notice I should also test?~~

Some tests don't pass on my local machine but they seem unrelated to anything I've written. For example the `analyze_once` test fails, because the error message isn't the expected one, or `commands.shard\hermetic\run_test.dart` bc it tries to access the flutter artifacts dir at `C:\C:\Users\hanne\devel\flutter_sdk\bin\cache\artifacts` which clearly isn't the right syntax. Maybe it's because I'm on Windows?

EDIT: Added all the tests. Edited schema URL.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

ardera committed:
add custom device support
- add CustomDevicesConfig class for getting the querying the currently configured device
- add CustomDeviceConfig which represents a specific device config
- make protocol discovery not assert a loopback address since the custom devices feature allows to not use port forwarding
- add a custom devices feature to enable custom devices support
- add string interpolation to utils.dart
- add some testing for custom devices

ardera committed:
- add license headers
- properly encode port forwarding success regex in json representation
- sort imports in device.dart
- make custom_device_workflow.dart null safe

ardera committed:
prefer relative imports

ardera committed:
- add tests, make non-null-safe
- add tests for custom device workflow
- add tests for custom device port forwarding
- add tests for custom device app starting / stopping
- add tests for custom device app installing / uninstalling
- split config into custom_devices_config.dart and custom_device_config.dart, so at least the CustomDeviceConfig can stay null-safe
- make CustomDevicesConfig non-null-safe since it relies on other non-null-safe stuff
- change JSON schema URI to refer to the stable branch
- fix other warnings
- use processManager.killPid instead of process.kill, since thats not compatible with testing

ardera committed:
- eliminate trailing U+0020 space character in custom_device_test.dart

ardera committed:
- add missing license header

ardera committed:
fix custom device workflow test

ardera commented:
In hindsight, I think it may be better to just define a `CustomApp` and some `custom_arm`, `custom_arm64`, `custom_x86`, `custom_x64` target platforms instead of reusing the linux ones, even if that may require additional support from any IDE plugins

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,725 @@
+      }
+    });
+
+    // if the process exists (even with exitCode == 0), that is considered
```
```suggestion
    // if the process exits (even with exitCode == 0), that is considered
```

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,725 @@
+      }
+    }));
+
+    // just some cleanup once the forwarding completed
```
nit: this comment isn't that useful. Avoid comments that just describe exactly what the code below is doing.

ardera replied:
fixed it

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,725 @@
+  @override
+  Future<void> dispose() async {
+    _sessions
+        ..forEach((_, CustomDeviceAppSession session) => session.dispose())
```
nit: 2 space indent, here and elsewhere

ardera replied:
fixed it

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,725 @@
+  }
+
+  @override
+  Future<bool> isAppInstalled(covariant ApplicationPackage app, {String userIdentifier}) {
```
mark the method as `async` then return `true/false` directly

ardera replied:
fixed it

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,725 @@
+        processManager: _processManager,
+        logger: _logger,
+      );
+    } else {
```
nit: remove the `else` since you have an early return above

ardera replied:
fixed it

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device_config.dart:
```diff
@@ -0,0 +1,144 @@
+        forwardPortSuccessRegex: explicitForwardPortSuccessRegex ? forwardPortSuccessRegex : (forwardPortSuccessRegex ?? this.forwardPortSuccessRegex)
+    );
+  }
+}
```
```suggestion
}

```

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart:
```diff
@@ -0,0 +1,90 @@
+  // it also may overwrite any things done by the user that aren't explicitly
+  // tracked by the JSON-representation. For example comments (not possible right now,
+  // but they'd be useful so maybe in the future) or formatting.
+}
```
```suggestion
}

```

jonahwilliams commented on packages/flutter_tools/lib/src/features.dart:
```diff
@@ -127,6 +133,8 @@ class FlutterFeatureFlags implements FeatureFlags {
     }
     return isEnabled;
   }
+
```
```suggestion
```

jonahwilliams commented on packages/flutter_tools/lib/src/features.dart:
```diff
@@ -296,6 +305,30 @@ const Feature flutterFuchsiaFeature = Feature(
+  name: 'alpha-quality support for custom, individually configured devices',
+  configSetting: 'enable-custom-devices',
+  environmentOverride: 'FLUTTER_CUSTOM_DEVICES',
+  extraHelpText: flutterNext ?
```
nit: no extra help text for this one

jonahwilliams commented on packages/flutter_tools/lib/src/features.dart:
```diff
@@ -296,6 +305,30 @@ const Feature flutterFuchsiaFeature = Feature(
 );
 
+const Feature flutterCustomDevicesFeature = Feature(
+  name: 'alpha-quality support for custom, individually configured devices',
```
I would leave out the branding designation. "Early support for custom device types"

jonahwilliams commented on packages/flutter_tools/lib/src/features.dart:
```diff
@@ -296,6 +305,30 @@ const Feature flutterFuchsiaFeature = Feature(
+    enabledByDefault: false,
+  ),
+  beta: FeatureChannelSetting(
+    available: flutterNext,
```
Lets keep this disabled for beta/stable initially

jonahwilliams commented on packages/flutter_tools/static/custom-devices.schema.json:
```diff
@@ -0,0 +1,111 @@
+      }
+    }
+  }
+}
```
```suggestion
}

```

jonahwilliams reviewed (commented):
Just took a first pass - overall looks good. We might need to think a bit about stabilization too. Not sure what your thoughts are there

ardera committed:
fix some issues
- formatting
- custom devices feature flag details
- set custom_devices.schema.json schema id to the stable one
- add "regex" format to the forwardPortSuccessRegex property in the schema

ardera committed:
Update packages/flutter_tools/lib/src/custom_devices/custom_device.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera commented:
okay, should all be resolved now.

> Just took a first pass - overall looks good. We might need to think a bit about stabilization too. Not sure what your thoughts are there

With stabilization, do you mean like API (or generally, interface) stability or making sure everything runs nice and stable?

ardera committed:
fix some documentation in the JSON schema

jonahwilliams commented:
Sort of both. I think we can say initially that we can change this at any version to give time to shake things out, and in the meantime keep it enabled only for master/dev. Any concerns with that?

jonahwilliams commented on packages/flutter_tools/lib/src/base/utils.dart:
````diff
@@ -390,3 +390,55 @@ bool isWhitespace(_AnsiRun run) {
+/// ```
+List<String> interpolateStringList(List<String> toInterpolate, Map<String, String> replacementValues) {
+  return toInterpolate.map((String s) => interpolateString(s, replacementValues)).toList();
+}
````
```suggestion
}

```

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart:
```diff
@@ -0,0 +1,90 @@
+
+/// Represents the custom devices config file on disk which in turn
+/// contains a list of individual custom device configs.
+class CustomDevicesConfig {
```
Can you walk me through how you are loading/locating this config ?

ardera replied:
Of course. Generally, it's heavily inspired `PersistentToolState`.

- `CustomDevicesConfig` is just stored in the context. It's specified as part of the global context in `context_runner.dart`.
- when `CustomDevicesConfig` is instantiated, it creates an internal `Config` object which on construction loads the config file from a path determined by the `Config` object (using the config name, which is `custom_devices.json`) and parses it.
- that path is `%APPDATA%/.flutter_<config name>` on Windows and `$HOME/.config/flutter/<config name>` on MacOS and linux (see `Config._configPath`)

However, now I'm not so sure my initial description about the config file being automatically created when the custom devices feature is enabled is correct. Will the context runner call all the fallback generators eagerly? Or only when they're first used?

jonahwilliams replied:
They're only called when it is first used - I'm not so sure about having side effects like this in the generators.

How does a user discover how to add this config?

ardera replied:
> They're only called when it is first used - I'm not so sure about having side effects like this in the generators.

Yeah, I agree.

> How does a user discover how to add this config?

You mean, like, how should they know where the config is located? 

We could create the config when `flutter config --enable-custom-devices` is called and additionally let it output some message like `The custom devices configuration can now be edited at "C:\..."` so they have a path they can edit. (So it's both no side effects in generators and discoverable)

jonahwilliams replied:
Is that what you are doing for your device? WDUT about having a special command to add a device configuration?

ardera replied:
> Is that what you are doing for your device?

Not sure what you mean, I think there's nothing about this that is special to my device. That single path that'd be output by `flutter config --enable-custom-devices` is the only path any custom device configuration would ever be sourced from.

> WDUT about having a special command to add a device configuration?

Yeah we can do that. Actually, in the initial description I mentioned that a new custom devices config file, when generated, contains the schema tag and a disabled example device config. The point of the latter is so people can just copy+paste that entry in the config file to add their own device.

But yeah, I think a command would be cool as well, especially so its usable by scripts etc. (Or do IDE plugins use the daemon for that? not sure)

jonahwilliams replied:
The daemon probably wouldn't support this command but it would definitely be usable from the CLI. Doesn't need to be done in this PR, but imagine something like:

`flutter custom-device add /path/to/confg`
`flutter custom-device list`
`flutter custom-device remove name`

ardera replied:
Okay, but why would multiple config files be useful? Right now it's just a single, centralised config files that contain all custom devices to be configured (also mentioned that in the issue iirc)

jonahwilliams replied:
I would imagine those commands as add a configuration to the existing config file, instead of editing a file in the SDK right?

ardera replied:
Yeah I mean if you just want to add some custom device, editing the SDK config is better IMO. You've got a structure and an example config you can start with (and autocompletion bc of the JSON schema but thats of course not relevant when you're not using VS Code). If you start with a blank separate config file instead, even I would probably get some names / structures wrong and would start looking at the SDK config for guidance anyway. Especially since you'll probably configure the custom devices once and then completely forget about it, so not enough "exposure" to remember all the details

But if want to automate it & make it able to be used by scripts (or generally, non-interactively), then something like `flutter custom-device add <config>` is definitely useful, I'm just not sure that's the use-case you're caring about here.

I think a command like this would be more useful instead:
```
flutter custom-device add --ssh --hostname raspberrypi --user pi --launch "flutter-pi --observatory-host 192.168.178.43" --no-use-port-forwarding
```

That's just my POV though. We could of course also do both and your proposed config command would look like `flutter custom-device add --config /path/to/config` in that case.

jonahwilliams replied:
Understood. I think the current strategy is fine for now FWIW, and we can address any usability concerns in the future.

ardera replied:
I'll resolve this for now, I'll create an issue to track this

ardera commented:
> Sort of both. I think we can say initially that we can change this at any version to give time to shake things out, and in the meantime keep it enabled only for master/dev. Any concerns with that?

Sure. Yeah it'd be nice to have release support too. Maybe when implementing that we'll have to make some other breaking changes.

I'm still probably going to create a fork and backport it to stable just so I can use it on stable myself

jonahwilliams commented:
It would help if we had a list of features that are incomplete, then maybe that could be the guidelines for a stable feature?

ardera committed:
Update packages/flutter_tools/lib/src/base/utils.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera commented:
> It would help if we had a list of features that are incomplete, then maybe that could be the guidelines for a stable feature?

Yep that's a good idea. So should we create a new GH project for that? Or a google doc?

jonahwilliams commented:
Either works for me, GH project might be a bit easier to manage though

ardera commented:
> Either works for me, GH project might be a bit easier to manage though

GH project sounds good to me. Can create some issues that you can add to the project (can't create the project myself of course). First one is [here](https://github.com/flutter/flutter/issues/78151) already.

> In hindsight, I think it may be better to just define a `CustomApp` and some `custom_arm`, `custom_arm64`, `custom_x86`, `custom_x64` target platforms instead of reusing the linux ones, even if that may require additional support from any IDE plugins

Created an issue for this: #78644 (Seems like you've already seen it). Though this doesn't need to be implemented in this PR since it's not blocking anything and can be easily changed later

ardera committed:
use custom platform type
- add `PlatformType.custom`
- make CustomDevices use `PlatformType.custom`
- make daemon report custom devices are always supported if custom devices feature is enabled

ardera commented:
- resolves #78644

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,724 @@
+
+    process.stdout.transform<String>(decoder)
+      .transform<String>(const LineSplitter())
+      .listen(
```
You can write this using a tear-off, here and elsewhere

`.listen(_logLinesController.add)`

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,724 @@
+
+    // if the process exits (even with exitCode == 0), that is considered
+    // a port forwarding failure and we complete with a null value.
+    unawaited(process.exitCode.then((int exitCode) {
```
nit: if you're not using the exit code use .whenComplete

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,724 @@
+      logger: _logger,
+      ipv6: ipv6,
+    );
+
```
uber-nit: the spacing here is a little odd, could you collect the related bits together?

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,724 @@
+  Future<bool> stop() async {
+    if (_process == null) {
+      return false;
+    } else {
```
don't need an `else` since you have an early return above

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,724 @@
+       _config = null,
+       super('custom devices');
+
+  CustomDevices.test({
```
If you can, mark this as `@visibleForTesting` and document any defaults

ardera replied:
I marked it `visibleForTesting` and documented it, did you mean that by documenting any defaults?

jonahwilliams replied:
yeah, sg

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart:
```diff
@@ -0,0 +1,90 @@
+         deleteFileOnFormatException: false
+       );
+
+  CustomDevicesConfig.test({
```
Make this `@visibleForTesting` if possible

ardera replied:
done

jonahwilliams reviewed (dismissed):
Took another pass, this looks really good so far

jonahwilliams commented:
My bad I didn't mean to approve yet, but its close!

ardera committed:
some style improvements
- use @visibleForTesting
- use whenComplete
- dont use `else` when the if returns unconditionally
- improve spacing
- add a comment explaining some order of things

ardera committed:
document CustomDevices constructors

ardera commented:
> My bad I didn't mean to approve yet, but its close!

nice! 😄

jonahwilliams commented on packages/flutter_tools/lib/src/context_runner.dart:
```diff
@@ -171,6 +172,11 @@ Future<T> runInContext<T>(
+        fileSystem: globals.fs,
+        logger: globals.logger,
+        platform: globals.platform,
+      )..ensureFileExists(),
```
rather that do this here (because it might be easy to miss), I would run this function in the constructor

ardera replied:
okay, fixed it

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,725 @@
+  @override
+  bool get canListAnything => _customDeviceWorkflow.canListDevices;
+
+  CustomDevicesConfig get customDevicesConfig => _config ?? context.get<CustomDevicesConfig>();
```
Avoid introducing `context` getters/setters. Instead ensure config is always provided

ardera replied:
Okay, yeah config is now not available in the context anymore and instead constructed when the `FlutterDeviceManager` constructer is called in `context_runner.dart:199`

jonahwilliams reviewed (commented):
I would love to see a more end to end test that creates a custom device config and verifies (using the fake process manager) that all of the correct hooks are called/ not called based on that config.

ardera commented:
While working on this, I've encountered a little gotcha with Windows' ping command: It won't always quit with exit code != 0 when the device is unreachable. It's also not that easy to detect error conditions with a regex since most of the output relevant is localized (internet search says you can search for the `ms` of the ping response time, but some other people said that doesn't work either, also not that easy if the hostname happens to contain a `ms` or something

Maybe we should let the SDK do the pinging and let the user specify a hostname / IP-Address that should be pinged, instead of a complete ping command?

jonahwilliams commented:
> Maybe we should let the SDK do the pinging and let the user specify a hostname / IP-Address that should be pinged, instead of a complete ping command?

How would moving the ping functionality into the tool help? If you don't want to use the windows ping, command you could always write a script that orchestrated pinging via Dart or some other language - and then define the script as a dart program

ardera commented:
> How would moving the ping functionality into the tool help? If you don't want to use the windows ping, command you could always write a script that orchestrated pinging via Dart or some other language - and then define the script as a dart program

Sure, I think that's better actually. Every windows user will have that problem so maybe we could include that script in the SDK? Or maybe keep it as a global pub package that gets activated when the custom devices feature is enabled?

jonahwilliams commented:
Keeping it as a pub package sounds reasonable, but it shouldn't be installed automatically. The goal of this feature is to support custom devices without extending the SDK beyond the basic interfaces

ardera commented:
I think I've found a way to make it work with a regex, I'll push that

ardera committed:
some fixes, improvements
- don't store custom devices config in context
- add optional ping success regex in config (by default will look for 'ms TTL=' in ping command output)
- add end-to-end test for custom devices
- make sure ports are unforwarded when app is terminated
- implicitly ensure the custom devices config file exists in the constructor rather than when invoking a separate method
- fix some formatting

ardera committed:
fix import for device port forwarding

ardera committed:
fix double indent

ardera commented:
> I would love to see a more end to end test that creates a custom device config and verifies (using the fake process manager) that all of the correct hooks are called/ not called based on that config.

Implemented it, it was a good idea, I actually found & fixed a bug that way!

ardera committed:
use own MockBundleBuilder in test instead of importing the one from another test

jonahwilliams commented:
Awesome! I will take another look in a bit.

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,736 @@
+      } else {
+        // null value means the forwarding failed (for whatever reason)
+        // increase port by one and try again
+        actualHostPort++;
```
nit: prefer `+=`

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,736 @@
+  @override
+  Future<void> dispose() async {
+    // copy the list so we don't modify it concurrently
+    List<ForwardedPort>.of(_forwardedPorts).forEach(unforward);
```
This needs to await all of the un-forwarding, something like:

```
await Future.wait([
  for (ForwardedPort port in _forwardedPorts)
    unforward(port)
]);
```

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,736 @@
+  List<ForwardedPort> get forwardedPorts => List<ForwardedPort>.unmodifiable(_forwardedPorts);
+
+  @override
+  Future<void> unforward(ForwardedPort forwardedPort) async {
```
Even though this method has no async now some could be added in the future (context for comment above on await unforwaard)

ardera replied:
Which method do you mean? the `forwardedPorts` property or `unforward`? Because `unforward` has an `async` right now

Btw seems like the documentation for `DevicePortForwarder.forwardedPorts` is wrong, it says it returns a Future of all the forwarded ports but it doesn't, just returns synchronously. (And the file name has a typo, the file is called `device_port_forwader.dart`)

jonahwilliams replied:
From L125

ardera replied:
but that has an `async` too, doesn't it?

jonahwilliams replied:
oops on the typo (FYI @jmagman )

At any rate, its important to await all of the futures you are using

jmagman replied:
🤦‍♀️  https://github.com/flutter/flutter/pull/79481

ardera replied:
ah, yeah I've implemented that; should all be awaited now

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_device.dart:
```diff
@@ -0,0 +1,736 @@
+  }
+}
+
+class CustomDeviceAppSession {
```
nit: please document this class with at least a sentence

jonahwilliams commented on packages/flutter_tools/lib/src/custom_devices/custom_devices_config.dart:
```diff
@@ -0,0 +1,97 @@
+  static const String _kCustomDevicesConfigName = 'custom_devices.json';
+  static const String _kCustomDevicesConfigKey = 'custom-devices';
+  static const String _kSchema = r'$schema';
+  static const String _defaultSchema = 'https://raw.githubusercontent.com/flutter/flutter/stable/packages/flutter_tools/static/custom-devices.schema.json';
```
I think this should actually refer to the schema via local path into the SDK instead of from github. Otherwise if we make changes locally or in a pre-release branch (dev/beta) users may see the incorrect configuration until that reaches stable.

You can use Cache.flutterRoot to find the root of the flutter checkout, then path manipulation from there.

ardera replied:
Yeah I see, a file path seems to be the better compromise.

jonahwilliams commented on packages/flutter_tools/test/general.shard/custom_devices/custom_device_test.dart:
```diff
@@ -0,0 +1,544 @@
+import '../../src/fake_process_manager.dart';
+import '../../src/fakes.dart';
+
+class MockBundleBuilder extends Mock implements BundleBuilder {}
```
Prefer using classes that extends `Fake` instead of adding more mocks. We're trying to avoid using new Mock classes to keep more null safety migration work from piling up.

jonahwilliams commented on packages/flutter_tools/test/general.shard/custom_devices/custom_device_test.dart:
```diff
@@ -0,0 +1,544 @@
+      ProcessManager: () => FakeProcessManager.any()
+    }
+  );
+}
```
```suggestion
}

```

jonahwilliams commented on packages/flutter_tools/test/general.shard/custom_devices/custom_device_workflow_test.dart:
```diff
@@ -0,0 +1,28 @@
+    expect(workflow.canListDevices, true);
+    expect(workflow.canListEmulators, false);
+  });
+}
```
```suggestion
}

```

jonahwilliams reviewed (commented):
Only big concern right now is schema location, everything else seems pretty good

ardera committed:
Update packages/flutter_tools/test/general.shard/custom_devices/custom_device_test.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
Update packages/flutter_tools/test/general.shard/custom_devices/custom_device_workflow_test.dart

Co-authored-by: Jonah Williams <jonahwilliams@google.com>

ardera committed:
add some more documentation
- create FakeBundleBuilder and use that in custom_device_test.dart instead of a MockBundleBuilder
- make default JSON schema for custom devices config a file URI, pointing to the schema inside the flutter install
- use += 1 instead of ++
- await unforwarding in CustomDevicePortForwarder.dispose

ardera committed:
prefer relative imports

jonahwilliams reviewed (approved):
LGTM

You'll need to resolve the merge conflicts - I moved most of the concrete DeviceDiscovery types out of devices.dart and into a separate library - you should do the same with the custom device code

ardera committed:
add CustomDevices discovery to flutter device manager again

ardera commented:
> LGTM
> 
> You'll need to resolve the merge conflicts - I moved most of the concrete DeviceDiscovery types out of devices.dart and into a separate library - you should do the same with the custom device code

Nice! fixed the conflicts

jonahwilliams commented:
Awesome, will land once tests pass and the build is green again
